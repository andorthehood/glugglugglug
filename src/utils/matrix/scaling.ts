export function scaling(sx: number, sy: number, sz: number) {
	return [sx, 0, 0, 0, 0, sy, 0, 0, 0, 0, sz, 0, 0, 0, 0, 1];
}
