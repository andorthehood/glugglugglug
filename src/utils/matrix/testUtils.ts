export function roundValue(value: number, precision = 6) {
	const factor = 10 ** precision;
	const rounded = Math.round(value * factor) / factor;
	return Object.is(rounded, -0) ? 0 : rounded;
}

export function formatMatrix(values: number[]) {
	return Array.from({ length: 4 }, (_, row) => {
		const start = row * 4;
		return values.slice(start, start + 4).map((value) => roundValue(value));
	});
}

export function formatVector(values: number[]) {
	return values.map((value) => roundValue(value));
}
