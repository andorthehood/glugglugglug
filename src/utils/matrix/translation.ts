export function translation(tx: number, ty: number, tz: number) {
	return [1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, tx, ty, tz, 1];
}
