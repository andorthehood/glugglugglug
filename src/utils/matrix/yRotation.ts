export function yRotation(angleInRadians: number) {
	const c = Math.cos(angleInRadians);
	const s = Math.sin(angleInRadians);

	return [c, 0, -s, 0, 0, 1, 0, 0, s, 0, c, 0, 0, 0, 0, 1];
}
