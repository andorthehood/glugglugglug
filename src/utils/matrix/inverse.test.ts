import { describe, expect, it } from "vitest";
import { inverse } from "./inverse";
import { multiply } from "./multiply";
import { scaling } from "./scaling";
import { translation } from "./translation";
import { yRotation } from "./yRotation";
import { formatMatrix } from "./testUtils";

describe("inverse", () => {
	it("inverts affine matrices", () => {
		const matrix = multiply(
			translation(4, -2, 7),
			multiply(yRotation(Math.PI / 6), scaling(2, 0.5, 3)),
		);
		const inv = inverse(matrix);
		const identity = multiply(matrix, inv, new Array<number>(16));

		expect(formatMatrix(matrix)).toMatchSnapshot();
		expect(formatMatrix(inv)).toMatchSnapshot();
		expect(formatMatrix(identity)).toMatchSnapshot();
	});
});
