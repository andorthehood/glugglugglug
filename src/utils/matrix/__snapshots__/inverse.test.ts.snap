// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`inverse > inverts affine matrices 1`] = `
[
  [
    1.732051,
    0,
    -1,
    0,
  ],
  [
    0,
    0.5,
    0,
    0,
  ],
  [
    1.5,
    0,
    2.598076,
    0,
  ],
  [
    4,
    -2,
    7,
    1,
  ],
]
`;

exports[`inverse > inverts affine matrices 2`] = `
[
  [
    0.433013,
    0,
    0.166667,
    0,
  ],
  [
    0,
    2,
    0,
    0,
  ],
  [
    -0.25,
    0,
    0.288675,
    0,
  ],
  [
    0.017949,
    4,
    -2.687393,
    1,
  ],
]
`;

exports[`inverse > inverts affine matrices 3`] = `
[
  [
    1,
    0,
    0,
    0,
  ],
  [
    0,
    1,
    0,
    0,
  ],
  [
    0,
    0,
    1,
    0,
  ],
  [
    0,
    0,
    0,
    1,
  ],
]
`;
